File: src/6. ZigZag Conversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */

function convert(s, numRows) {
  if (numRows <= 1) return s
  let zigzag = new Array(numRows).fill('')
  let currRow = 0,
    goDown = true
  for (let char of s) {
    zigzag[currRow] += char
    if (goDown) {
      if (currRow < numRows - 1) currRow++
      else {
        goDown = false
        currRow--
      }
    } else {
      if (currRow > 0) currRow--
      else {
        goDown = true
        currRow++
      }
    }
  }
  return zigzag.join('')
}

File: src/6. Zigzag Conversion.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows <= 1) {
            return s;
        }

        vector<string> zigzag(numRows);
        bool goDown = true;
        int currRow = 0;
        for (char c : s) {
            zigzag[currRow] += c;
            if (goDown && currRow != numRows - 1) {
                currRow++;
            } else if (goDown && currRow == numRows - 1) {
                goDown = false;
                currRow--;
            } else if (!goDown && currRow != 0) {
                currRow--;
            } else {
                goDown = true;
                currRow++;
            }
        }

        string ans;
        for (string z : zigzag) {
            ans += z;
        }
        return ans;
    }
};